Implement a receiver for MicroG SQLite databases, without root.

tlambertz opened the pull request:
This pull request allows any other app to share a MicroG SQLite exposure database file with this app, which then automatically analyzes it. In combination with https://github.com/microg/GmsCore/pull/1291 it allows for root-less analysis of the database.

This is done by registering this app as a receiver for any `intent.action.SEND` with mime-type `application/vnd.sqlite3`. In `onCreate`, it is checked whether a database was received this way, and the mode of the app is temporarily switched to MicroG. The analysis is then done on this passed file. Since we can only access it as a file-descriptor, and SQLite expects a file, it is temporarily stored in the cache as a file.

Is there a reason you used the `singleTask` launch mode? This would complicate implementation a bit, I replaced it with `standard`. 

It might be nice to have another file-picker like solution for all app modes, but I was unsure how to best implement that. The current solution has a really nice workflow for me:
- open microg, click export, select companion app, analysis starts automatically.

mh- commented:
I added `singleTask` as a quick fix some time ago, to avoid multiple background matching threads being started in parallel (and when they try to finish, crash the app). Will have to look into possible side-effects of going back to `standard`.

mh- commented:
Ok, now I remember what goes wrong with `launchMode="standard"`: 
While matching is running, tap on "About this app", then tap on the left arrow in the top left corner.
With `launchMode="singleTask"`, this works.
But with `launchMode="standard"`, nothing happens for a short moment, then the screen goes black.

corneliusroemer commented:
@tlambertz  Is there a way this could be extended to work with the CCTG app that ships with the MicroG ENF-implementation - without requiring explicit MicroG installation?

Presumably the CCTG folks would have to do the significant part of the implementation for this.

Maybe I can try to get a workaround going using adb meanwhile.

tlambertz commented:
Hm I can reproduce the error you encounter without single-task mode. Though that seems like a dirty workaround. I am not a well versed android developer, but I think the core issue is that that database loading should not happen in `onCreate`.

The app is unable to show anything while the `onCreate` function is running. Loading the database is a potentially longer running task, so should be moved to a worker thread?

(In my implementation i wanted to show a toast `Switching App to MicroG Mode`, but that did not work well, since toasts are not rendered until `onCreate` is done)

The real issue I have with single-task mode in the context of this pull request is a different one though: `onCreate()` does NOT get called again when another app exports the database and the companion is already running. I think I can workaround this by using `onNewIntent`, which might be the quickest way to get this working.
I found https://inthecheesefactory.com/blog/understand-android-activity-launchmode/en to be a good resource in understanding the launch-modes.

ale5000-git commented:
It got merged, but using: `application/vnd.microg.exposure+sqlite3`

tlambertz commented:
I have rebased and changed the mime-type.

I have also reverted to using singleTask and use `onNewIntent` to recreate the app. The user-experience is not the best, since with a full 14day database the app is stuck in a black screen for ~10-20 seconds, while nothing can be shown since `onCreate()` is blocking. Except for that everything works fine together with the MicroG master branch.

mh- reviewed (approved):
Thank you!

ale5000-git commented:
Just a minor thing: The correct name is `microG` but you have `MicroG` in log messages in `microg` in comments.

mh- commented:
@ale5000-git don't worry, we'll fix this
